### ISSUE TITLE
[web-test-runner] Conversion of v8 test data to istanbul is slow

### ISSUE THREAD

**Issue by benjamind** [modernweb-dev/web#2023]
We've discovered that there is significant overhead in the conversion of code coverage reports by WTR from the v8 format output from chrome dev tools protocol, to the istanbul format used in the reporters.

In projects with larger code graphs under test this conversion time is upwards of 30s on fast machines, much worse in CI.

From digging around the only reason I can conclude for this conversion is the compatibility with other coverage reporting tools that use istanbul. Primarily this seems to be the generation of a HTML coverage report and the JUnit output.

Would the team consider moving away from istanbul to speed up the test run? This might entail writing a custom HTML report using the v8 data, but could be a significant speed up?

**Comment by daKmoR** [comment]
interesting find... are you running your tests in chrome only?
I'm unsure how that would work on other browsers? or via selenium?

I think a flag could be an option...

@LarsDenBakker what's your though on this?

**Comment by benjamind** [comment]
Yes this is in chrome only. Well chrome for coverage. We separately run Firefox tests also. We're running with the playwright test runner.

I agree it wouldn't work with selenium or other browsers of course. Not sure the best path towards this, perhaps coupling a specific config of the playwright runner to generate v8 only with a plugin for a new reporter that takes that format?

**Comment by 43081j** [comment]
i think i figured most of it out now, got our test run down from ~15 minutes to ~6 minutes with the following:

#### Parallel computation (minor)

[this loop](https://github.com/modernweb-dev/web/blob/1eb5345267e59fbcd6054064a4ccdb412b826711/packages/test-runner-coverage-v8/src/index.ts#L47) runs the async tasks in serial. we could improve this to be like a `Promise.all` (but can't do a `Promise.all` because it makes WDS sad, sending so many requests at once. some kind of parallelism limit could work)

**perf improvement:** minor

#### Only fetch matching sourcemaps

[here](https://github.com/modernweb-dev/web/blob/1eb5345267e59fbcd6054064a4ccdb412b826711/packages/test-runner-coverage-v8/src/index.ts#L63-L69) we unconditionally fetch the sourcemap of a given file.

later, we decide if we even want to compute coverage for this file or not.

we should move that line into the condition (it isn't needed any earlier anyway). 

**perf improvement:** definitely noticeable

#### Load the file into coverage _once_

[here](https://github.com/modernweb-dev/web/blob/1eb5345267e59fbcd6054064a4ccdb412b826711/packages/test-runner-coverage-v8/src/index.ts#L74-L75) we load the given file on disk into the v8-to-istanbul function.

this results in it actually reading the file, loading it, etc.

within one page, we don't see much duplication here since its unlikely v8 would give us dupes in the coverage array.

however, across multiple calls to `v8ToIstanbul`, we _regularly_ pass the same files in (pages have imports in common). if we can cache the converter like this:

```ts
                  let converter = ccache.get(filePath);
                  if (!converter) {
                    converter = v8_to_istanbul_1.default(filePath, 0, sources);
                    await converter.load();
                  }
                  ccache.set(filePath, converter);
```

we only load the file off disk once, and reuse it many times.

**perf improvement:** massive, several minutes cut off our test runs

---

the last is the biggest, but off the top of my head im not sure where such a cache could live. as it'd need to be reused across many calls to this v8toistanbul function. 

maybe we give it an optional cache param or something, a map, which it can reuse as state. not sure

cc @Westbrook

**Comment by 43081j** [comment]
@Westbrook weirdly using an lrucache with a `max: 200` took it down to 4 minutes rather than 6!

witchcraft

seems like a good option then. i suppose we should be careful with how we key it though

im currently keying by the browser url, which seems fine. but maybe across multiple `v8ToIstanbul` calls, two browser urls could mean different things?
